[Proposal] Azure VM - Instance View Filter

stefangordon opened the pull request:
So we have several situations where there are common fields not accessible on a resource during enumeration, but we'd hate to go do a detail enumeration to fill them in if the user isn't filtering on them.

In this case the InstanceView fields can't be expanded during enumeration, but contain the indicator of whether the VM is actually turned on.

Thoughts on this proposal of having a value filter populate a detail field on demand? 

Note that the resulting dictionary actually matches the schema you'd get if you individually queried a VM and asked to have this data included, so the json paths make _sense_.

Example:

```
 policies:
   - name: running-vm
     resource: azure.vm
     filters:
      - type: instance-view
        key: statuses[].code
        op: in
        value_type: swap
        value: "PowerState/running"
```

(resolves #2254)

anastasiia-zolochevska commented on tools/c7n_azure/c7n_azure/resources/vm.py:
```diff
@@ -23,4 +24,21 @@ class resource_type(object):
+class InstanceViewFilter(ValueFilter):
+    schema = type_schema('instance-view', rinherit=ValueFilter.schema)
+
+    def __call__(self, i):
```
We should keep the practice of having an example of the policy in comments

stefangordon replied:
Ah good catch, I'll add unit tests and docs on this too if we go ahead with it.

kapilt commented on tools/c7n_azure/c7n_azure/resources/vm.py:
```diff
@@ -23,4 +24,21 @@ class resource_type(object):
+    def __call__(self, i):
+        if 'instanceView' not in i:
+            client = self.manager.get_client()
+            instance = client.virtual_machines.get(i['id'].split('/')[4], i['name'], expand='instanceview').instance_view
```
might be worth getting a utility function for this

stefangordon replied:
👍 Yeah @anastasiia-zolochevska was actually thinking about overriding `augment` for ARM resources to set a `resourceGroup` field.  Sort of the most common field in the whole resource, will be good to have it readily accessible.  (or put some utility method in and a custom filter)

stefangordon replied:
Updated to use resourcegroup that has since been augmented into all ARM objects

kapilt commented on tools/c7n_azure/c7n_azure/resources/vm.py:
```diff
@@ -23,4 +24,21 @@ class resource_type(object):
+        if 'instanceView' not in i:
+            client = self.manager.get_client()
+            instance = client.virtual_machines.get(i['id'].split('/')[4], i['name'], expand='instanceview').instance_view
+            i['instanceView'] = instance.serialize()
```
we typically try to annotate those with a prefix, so that the serialized representation for object storage/disk is canonical with markups from additional lookups clearly annotated. also given the usage/context  we would typically call the super in this context with the instanceview instead of the instance so the user can specify matches with root of keys local to the instance view.

stefangordon replied:
So in this case I suspected this might be ideal, because this is actually the schema you would get if you did a `get` on a VM.  It is just the specific enumerator we use that omits this field for performance reasons (although the SDK is working on fixing it).

kapilt commented on tools/c7n_azure/c7n_azure/resources/vm.py:
```diff
@@ -23,4 +24,21 @@ class resource_type(object):
-        name = 'name'
\ No newline at end of file
+        name = 'name'
+        default_report_fields = (
```
yay report fields, i'm hopeful that means custodian report now works here.

stefangordon replied:
Yeah, this was one thing that was making us think of augmenting the `resourceGroup` in as a pre-parsed field instead of a custom filter - because it would always be a report field you'd want.

kapilt commented on tools/c7n_azure/c7n_azure/resources/vm.py:
```diff
@@ -23,4 +24,21 @@ class resource_type(object):
+@VirtualMachine.filter_registry.register('instance-view')
+class InstanceViewFilter(ValueFilter):
+    schema = type_schema('instance-view', rinherit=ValueFilter.schema)
+
```
so we've been annotating IAM permissions typically for filters and actions, so that we can gen access/rbac policies for custodian execution based on a user's given policies (ie minimal permission set based on things to do), i'm not familiar with azure iam/rbac models but something to keep in mind.

stefangordon replied:
I saw that was happening and haven't had a chance to work through what it means for us.   Filed #2273

stefangordon commented:
Great feedback, will be iterating on this a bit and push updates, thanks.

stefangordon commented:
Updates based on feedback, added unit test (and fixed a test decorator bug), added a use case to the documentation.  Used heavily this morning for some production subscription work, works great.

aluong commented on tools/c7n_azure/c7n_azure/resources/vm.py:
```diff
@@ -31,3 +31,15 @@ class resource_type(object):
+    schema = type_schema('instance-view', rinherit=ValueFilter.schema)
+
+    def __call__(self, i):
+        if 'instanceView' not in i:
```
I would prefer a more descriptive variable name here than "i".

stefangordon replied:
I think `i` is used in every filter implementation (it's an interator and `i` is the current item)
